Framework/Algorithms/src/GetEiMonDet3.cpp

```diff
@@ -0,0 +1,305 @@
+  const double energy = 0.5 * NeutronMass * velocity * velocity / meV;
+  progress(1.0);
+  g_log.notice() << "Final time-of-flight:" << timeOfFlight << " which gives "
+                 << energy << " as " << Prop::INCIDENT_ENERGY << ".\n";
+  setProperty(Prop::INCIDENT_ENERGY, energy);
+}
+
+/** Calculates the time of flight from the monitor to the detectors.
+ *
+ *  Adds pulse intervals to the TOF until it is greater than `minTOF`.
+ *  @param detectorWs Detector workspace
+ *  @param detectorEPP The position of the detectors' elastic peak
+ *  @param monitorEPP The position of the monitor's elastic peak
+ *  @param minTOF minimum expected time of flight, in microseconds
+ *  @return The time of flight between the monitor and the detectors
+ */
+double GetEiMonDet3::computeTOF(const API::MatrixWorkspace &detectorWs,
+                                const double detectorEPP,
+                                const double monitorEPP, const double minTOF) {
+  double timeOfFlight = detectorEPP - monitorEPP;
+  // Check if the obtained time-of-flight makes any sense.
+  while (timeOfFlight <= minTOF) {
+    double pulseInterval = getProperty(Prop::PULSE_INTERVAL);
+    if (pulseInterval == EMPTY_DBL()) {
+      if (detectorWs.run().hasProperty(SampleLogs::PULSE_INTERVAL)) {
+        pulseInterval = detectorWs.run().getPropertyAsSingleValue(
+            SampleLogs::PULSE_INTERVAL);
+        pulseInterval *= 1e6; // To microseconds.
+      } else {
+        throw std::invalid_argument(
+            "PulseInterval not explicitly given nor found in the sample logs.");
+      }
+    }
+    g_log.notice() << "Frame delay of " << pulseInterval
+                   << " microseconds will be added to the time-of-flight.\n";
+    timeOfFlight += pulseInterval;
+  }
+  g_log.notice() << "Calculated time-of-flight: " << timeOfFlight << ".\n";
+  return timeOfFlight;
+}
+
+/** Runs GroupDetectors on given workspace indices
+ *
+ * @param ws a workspace to group
+ * @param wsIndices a vector of workspace indices to group
+ * @return a single spectrum workspace
+ */
+API::MatrixWorkspace_sptr
+GetEiMonDet3::groupSpectra(API::MatrixWorkspace_sptr &ws,
+                           const std::vector<size_t> &wsIndices) {
+  auto group = createChildAlgorithm("GroupDetectors");
+  group->setProperty("InputWorkspace", ws);
+  group->setProperty("OutputWorkspace", "unused");
+  group->setProperty("WorkspaceIndexList", wsIndices);
+  group->execute();
+  return group->getProperty("OutputWorkspace");
+}
+
+/** Computes the minimum TOF between monitor and detectors from maximum
+ *  energy
+ *
+ * @param ws a workspace containing the instrument
+ * @param sampleToDetectorDistance the l2 distance
+ * @return minimum expected TOF, in microseconds
+ */
+double GetEiMonDet3::minimumTOF(const API::MatrixWorkspace &ws,
+                                const double sampleToDetectorDistance) {
+  const double minEnergy = getProperty(Prop::MAX_ENERGY);
```

Comment:
would this make more sense being called maxEnergy or am I missing something?

Reply:
It should read `maxEnergy`, indeed. I think I got confused with the inverse proportionality of energy and TOF.